[WIP] adding GenreSearchlight correlation script

jamalw opened the pull request.

buildbot-princeton commented:
Can one of the admins verify this patch?

mihaic commented:
Jenkins, add to whitelist.

mihaic commented:
PSA: @mjanderson09 and @jamalw are still working on this, so hold on with any reviews.

mjanderson09 commented:
Right, I think what's needed at this point is to:

1.) Add mask file to examples/searchlight directory
2.) Add option for generating random data, in addition to loading in real data from files
3.) Fix any style issues so it passes the pr-check.sh script

jamalw commented:
Hi Michael,
So I just added the full brain mask to the searchlight examples directory.
Working on steps 2 and 3 right now. Also, Ken said that it is okay for you
have a run from an actual subject as long as you anonymize it by distorting
the face of the subject.

- Jamal


On Mon, Apr 3, 2017 at 12:48 PM, Michael Anderson <notifications@github.com>
wrote:

> Right, I think what's needed at this point is to:
>
> 1.) Add mask file to examples/searchlight directory
> 2.) Add option for generating random data, in addition to loading in real
> data from files
> 3.) Fix any style issues so it passes the pr-check.sh script
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-291202408>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2RqcD2S4kq463ohbkiYZDKUx_xwA37ks5rsSL2gaJpZM4Mv9jJ>
> .
>

mjanderson09 commented:
Hi Jamal,

Awesome! Thank you. The only issue then with real data is the size (9GB subject data is too big). I think even a few TRs would be too much. So maybe random data is good, but then we could potentially provide a link to some cloud storage with a small version of the dataset.

Thanks!
Mike

jamalw commented:
I see. No problem. I should have the random data set generator added by
today!

- Jamal


On Mon, Apr 3, 2017 at 1:06 PM, Michael Anderson <notifications@github.com>
wrote:

> Hi Jamal,
>
> Awesome! Thank you. The only issue then with real data is the size (9GB
> subject data is too big). I think even a few TRs would be too much. So
> maybe random data is good, but then we could potentially provide a link to
> some cloud storage with a small version of the dataset.
>
> Thanks!
> Mike
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-291207679>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2RqfMkHN8JFPEGrUdjzNr62lCmsavpks5rsScbgaJpZM4Mv9jJ>
> .
>

mihaic commented:
@mjanderson09, we are waiting for your review.

mjanderson09 commented:
OK, Great! This looks good. Three small things I noticed:

Line 4, should it be "from brainiak.searchlight" not "from brainiak.brainiak.searchlight". On my machine it worked for the first one when I was running from the examples directory.

Line 8, there is an absolute path on your machine. I would change it to either './' which assumes they are running in the examples directory, or take the path as an argument to the script

Line 24 should be += instead of just =, must be a typo.

Thanks!
Mike

mihaic commented:
Actually, @jamalw, skip the "./". Simply leave `datadir` blank, which also represents the current directory.

jamalw commented:
No problem Mihai. I've deleted datadir from the script. Mike, I originally
called brainiak twice because of a weird installation issue i had. To fix
this, I moved the brainiak directory to my local working directory. Also, I
went ahead and fixed the typo on line 24. The newest version of the script
has been added to github. Let me know what you think!

- Jamal


On Wed, Apr 19, 2017 at 3:01 PM, Mihai Capotă <notifications@github.com>
wrote:

> Actually, @jamalw <https://github.com/jamalw>, skip the "./". Simply
> leave datadir blank, which also represents the current directory.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-295392802>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2RqZdQFtuEUFrqxTVJ3rUgeSka9q6zks5rxloUgaJpZM4Mv9jJ>
> .
>

jamalw commented:
Thanks!

- Jamal


On Wed, Apr 19, 2017 at 4:04 PM, Michael Anderson <notifications@github.com>
wrote:

> *@mjanderson09* approved this pull request.
>
> Thank you for addressing the issues. This looks good!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#pullrequestreview-33605242>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2RqVx94y5hVlAYAquCMFRq9r91sTG1ks5rxmjqgaJpZM4Mv9jJ>
> .
>

mihaic commented on examples/searchlight/GenreSearchlight.py:
```diff
@@ -0,0 +1,75 @@
+import numpy as np
```
Please add a paragraph or two describing what the example does.

mihaic replied:
The file should be named like any other Python module, e.g., "genre_searchlight_example.py":
https://www.python.org/dev/peps/pep-0008/#naming-conventions

mihaic commented on examples/searchlight/GenreSearchlight.py:
```diff
@@ -0,0 +1,75 @@
+from scipy import stats
+from scipy.sparse import random
+
+np.seterr(divide='ignore',invalid='ignore')
```
Add a comment explaining why you ignore these errors.

mihaic commented on examples/searchlight/GenreSearchlight.py:
```diff
@@ -0,0 +1,75 @@
+ d1_reshape[a1_idx[0],i] += classical
+ d1_reshape[a1_idx[0],i+8] += jazz
+ d2_reshape[a1_idx[0],i] += classical
+ d2_reshape[a1_idx[0],i+8] += jazz
```
Use 4 spaces per indentation level:
https://www.python.org/dev/peps/pep-0008/#indentation

mihaic commented on examples/searchlight/GenreSearchlight.py:
```diff
@@ -0,0 +1,75 @@
+    classical_within  = np.mean(corrAB[0:8,0:8])
+    jazz_within       = np.mean(corrAB[8:16,8:16])
+    classJazz_between = np.mean(corrAB[8:16,0:8])
+    jazzClass_between = np.mean(corrAB[0:8,8:16])
```
Do not align assignments:
https://www.python.org/dev/peps/pep-0008/#pet-peeves

jamalw commented:
Hey Mihai,
I made all of the changes except making the nifti smaller. The issue is
that I can't downsample it to 100KB to my knowledge. The lowest I could
downsample it is in the MB range but this would involve changing the
dimensions of the synthetic brain as well. If I could downsample it to
something lower I'm not sure how.

Last thing to note is that the command for the errors that were getting
dismissed is no longer necessary because it was used at an earlier stage of
code development. So I deleted that command.

The newest script is on github!

- Jamal


On Wed, Apr 19, 2017 at 5:45 PM, Mihai Capotă <notifications@github.com>
wrote:

> *@mihaic* requested changes on this pull request.
>
> I pointed out in the comments some more minor changes you should do before
> we merge the code.
>
> Additionally, it would be great if you could use a smaller Nifti file (say
> about 100 kB).
>
> Finally, remove "examples/searchlight/searchlight_images/.DS_Store".
> ------------------------------
>
> In examples/searchlight/GenreSearchlight.py
> <https://github.com/IntelPNI/brainiak/pull/204#discussion_r112319315>:
>
> > @@ -0,0 +1,75 @@
> +import numpy as np
>
> Please add a paragraph or two describing what the example does.
> ------------------------------
>
> In examples/searchlight/GenreSearchlight.py
> <https://github.com/IntelPNI/brainiak/pull/204#discussion_r112319599>:
>
> > @@ -0,0 +1,75 @@
> +import numpy as np
>
> The file should be named like any other Python module, e.g.,
> "genre_searchlight_example.py":
> https://www.python.org/dev/peps/pep-0008/#naming-conventions
> ------------------------------
>
> In examples/searchlight/GenreSearchlight.py
> <https://github.com/IntelPNI/brainiak/pull/204#discussion_r112319742>:
>
> > @@ -0,0 +1,75 @@
> +import numpy as np
> +from nilearn.image import load_img
> +import sys
> +from brainiak.searchlight.searchlight import Searchlight
> +from scipy import stats
> +from scipy.sparse import random
> +
> +np.seterr(divide='ignore',invalid='ignore')
>
> Add a comment explaining why you ignore these errors.
> ------------------------------
>
> In examples/searchlight/GenreSearchlight.py
> <https://github.com/IntelPNI/brainiak/pull/204#discussion_r112320025>:
>
> > +
> +# Generate random data
> +data1_rand = np.random.rand(91,109,91,16)
> +data2_rand = np.random.rand(91,109,91,16)
> +classical = np.random.rand(2600)
> +jazz = np.random.rand(2600)
> +d1_reshape = np.reshape(data1_rand,(91*109*91,16))
> +d2_reshape = np.reshape(data2_rand,(91*109*91,16))
> +a1 = load_img('a1plus_2mm.nii')
> +a1_vec = np.reshape(a1.get_data(),(91*109*91))
> +a1_idx = np.nonzero(a1_vec)
> +for i in range(8):
> + d1_reshape[a1_idx[0],i] += classical
> + d1_reshape[a1_idx[0],i+8] += jazz
> + d2_reshape[a1_idx[0],i] += classical
> + d2_reshape[a1_idx[0],i+8] += jazz
>
> Use 4 spaces per indentation level:
> https://www.python.org/dev/peps/pep-0008/#indentation
> ------------------------------
>
> In examples/searchlight/GenreSearchlight.py
> <https://github.com/IntelPNI/brainiak/pull/204#discussion_r112320287>:
>
> > +# Load mask
> +mask_img = load_img('MNI152_T1_2mm_brain_mask.nii')
> +mask_img = mask_img.get_data()
> +
> +# Definte function that takes the difference between within vs. between genre comparisons
> +def corr2_coeff(AB,msk,myrad,bcast_var):
> +    if not np.all(msk):
> +        return None
> +    A,B = (AB[0], AB[1])
> +    A = A.reshape((-1,A.shape[-1]))
> +    B = B.reshape((-1,B.shape[-1]))
> +    corrAB = np.corrcoef(A.T,B.T)[16:,:16]
> +    classical_within  = np.mean(corrAB[0:8,0:8])
> +    jazz_within       = np.mean(corrAB[8:16,8:16])
> +    classJazz_between = np.mean(corrAB[8:16,0:8])
> +    jazzClass_between = np.mean(corrAB[0:8,8:16])
>
> Do not align assignments:
> https://www.python.org/dev/peps/pep-0008/#pet-peeves
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#pullrequestreview-33625707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2Rqd1kLk36x7uARC5936MqcjxPAIlmks5rxoB4gaJpZM4Mv9jJ>
> .
>

mihaic commented:
@jamalw, you forgot to remove "examples/searchlight/searchlight_images/.DS_Store". Also, given that you created a new file instead of renaming the existing one, you should remove "examples/searchlight/GenreSearchlight.py".

Let's see if anyone has an idea about reducing the size of the Niftin file in the weekly meeting tomorrow.

jamalw commented:
I removed it from my local repo but I can't seem to find my brainiak branch
on github. Do you have any suggestions for finding it?

- Jamal


On Thu, Apr 20, 2017 at 4:20 PM, Mihai Capotă <notifications@github.com>
wrote:

> @jamalw <https://github.com/jamalw>, you forgot to remove
> "examples/searchlight/searchlight_images/.DS_Store". Also, given that you
> created a new file instead of renaming the existing one, you should remove
> "examples/searchlight/GenreSearchlight.py".
>
> Let's see if anyone has an idea about reducing the size of the Niftin file
> in the weekly meeting tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-295891594>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2RqX-e6aSb_IKJnWOK0Ti_NtNJN7Z-ks5rx74WgaJpZM4Mv9jJ>
> .
>

mihaic commented:
@jamalw, let's do Git debugging somewhere else. Email me or contact me on Gitter.

jamalw commented:
Sounds good.

On Thu, Apr 20, 2017 at 4:49 PM Mihai Capotă <notifications@github.com>
wrote:

> @jamalw <https://github.com/jamalw>, let's do Git debugging somewhere
> else. Email me or contact me on Gitter.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-295906249>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2RqZz4dYI2pFVsj1akl5-vnLuT65gaks5rx8TXgaJpZM4Mv9jJ>
> .
>
-- 
- Jamal

mihaic commented:
@jamalw, @kennethanorman offered to help with reducing the size of the Niftifile. Please get in touch and update the PR based on the result.

jamalw commented:
I just zipped the file and now it's 5kb instead of 3.6 MB. Would this
suffice Mihai?

- Jamal


On Fri, Apr 21, 2017 at 2:02 PM, Mihai Capotă <notifications@github.com>
wrote:

> @jamalw <https://github.com/jamalw>, @kennethanorman
> <https://github.com/kennethanorman> offered to help with reducing the
> size of the Niftifile. Please get in touch and update the PR based on the
> result.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-296261647>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2RqeiyAvv3aN_axbSO4rNSteD2iDpmks5ryO8kgaJpZM4Mv9jJ>
> .
>

mihaic commented:
Perfect!

jamalw commented:
Cool, I'm pushing it now.

- Jamal


On Fri, Apr 21, 2017 at 3:02 PM, Mihai Capotă <notifications@github.com>
wrote:

> Perfect!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-296278744>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2RqV0kFT7l10ofyYGTEubvDXwH8wk_ks5ryP0pgaJpZM4Mv9jJ>
> .
>

jamalw commented:
Just pushed it!

- Jamal


On Fri, Apr 21, 2017 at 3:02 PM, Jamal Williams <jamalawilliams@gmail.com>
wrote:

> Cool, I'm pushing it now.
>
> - Jamal
>
>
> On Fri, Apr 21, 2017 at 3:02 PM, Mihai Capotă <notifications@github.com>
> wrote:
>
>> Perfect!
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-296278744>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AE2RqV0kFT7l10ofyYGTEubvDXwH8wk_ks5ryP0pgaJpZM4Mv9jJ>
>> .
>>
>
>

mihaic commented:
@jamalw, it looks like you are in control of Git again. Please remove "examples/searchlight/searchlight_images/.DS_Store" and "examples/searchlight/GenreSearchlight.py".

Furthermore, you should remove the non-compressed Nifti file. Also, you should change the code to access the compressed one instead; I have not seen Nibabel reading zip files, so perhaps it is better to create a ".nii.gz" file instead.

mihaic commented on examples/searchlight/genre_searchlight_example.py:
```diff
@@ -0,0 +1,75 @@
+for (cnt, img) in enumerate(global_outputs):
+    plt.imshow(img,vmin=minval,vmax=maxval)
+    plt.colorbar()
+    plt.savefig('searchlight_images/' + 'img' + str(cnt) + '.png')
```
This generates an error because the "searchlight_images" directory does not exist. Either create it, or save the image in the current directory.

jamalw commented:
So i should put the searchlight_images folder back in github? I thought you
wanted me to delete it. I created it so all of the images wouldn't pile up
in the working directory.

- Jamal


On Fri, Apr 21, 2017 at 6:43 PM, Mihai Capotă <notifications@github.com>
wrote:

> *@mihaic* commented on this pull request.
> ------------------------------
>
> In examples/searchlight/genre_searchlight_example.py
> <https://github.com/IntelPNI/brainiak/pull/204#discussion_r112789296>:
>
> > +sl.distribute([data1,data2],mask_img)
> +sl.broadcast(None)
> +print('Running Searchlight...')
> +global_outputs = sl.run_searchlight(corr2_coeff)
> +
> +# Plot searchlight results
> +maxval = np.max(global_outputs[np.not_equal(global_outputs,None)])
> +minval = np.min(global_outputs[np.not_equal(global_outputs,None)])
> +global_outputs = np.array(global_outputs, dtype=np.float)
> +print(global_outputs)
> +
> +import matplotlib.pyplot as plt
> +for (cnt, img) in enumerate(global_outputs):
> +    plt.imshow(img,vmin=minval,vmax=maxval)
> +    plt.colorbar()
> +    plt.savefig('searchlight_images/' + 'img' + str(cnt) + '.png')
>
> This generates an error because the "searchlight_images" directory does
> not exist. Either create it, or save the image in the current directory.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#pullrequestreview-34136618>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2Rqf5Zle7X-gG1SOGCuD01aRqfGacrks5ryTEvgaJpZM4Mv9jJ>
> .
>

mihaic commented:
@jamalw, I wanted you to delete "examples/searchlight/searchlight_images/.DS_Store", not the directory itself. I was suggesting you create the directory at runtime, from you example script, if it does not exist (e.g., created by a previous run of the script).

jamalw commented:
Ah I see. I'll make that change now.

On Fri, Apr 21, 2017 at 6:48 PM Mihai Capotă <notifications@github.com>
wrote:

> @jamalw <https://github.com/jamalw>, I wanted you to delete
> "examples/searchlight/searchlight_images/.DS_Store", not the directory
> itself. I was suggesting you create the directory at runtime, from you
> example script, if it does not exist (e.g., created by a previous run of
> the script).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-296322591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2RqbvyGRDRA9ZmqXyMzq5u31pmgpuUks5ryTJCgaJpZM4Mv9jJ>
> .
>
-- 
- Jamal

mihaic commented on examples/searchlight/genre_searchlight_example.py:
```diff
@@ -0,0 +1,80 @@
+print(global_outputs)
+
+# Save searchlight images
+dir = "searchlight_images"
```
Please use a name that is not used by existing Python objects:
https://docs.python.org/3/library/functions.html#dir

You could use `out_dir`.

mihaic commented:
@jamalw, thank you for your contribution.

Note that I updated your branch with the latest changes from the master branch. Before making further changes on your computer, make sure to pull the changes from the master branch of your GitHub fork and then from the master branch of IntelPNI.

For the future, make a new branch every time you start working on a new project instead of changing the master branch.

jamalw commented:
Got it. Thanks for all of the help Mihai!

- Jamal


On Fri, Apr 21, 2017 at 8:05 PM, Mihai Capotă <notifications@github.com>
wrote:

> @jamalw <https://github.com/jamalw>, thank you for your contribution.
>
> Note that I updated your branch with the latest changes from the master
> branch. Before making further changes on your computer, make sure to pull
> the changes from the master branch of your GitHub fork and then from the
> master branch of IntelPNI.
>
> For the future, make a new branch every time you start working on a new
> project instead of changing the master branch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/IntelPNI/brainiak/pull/204#issuecomment-296331468>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2Rqb-DAJGPoEoaTliDtWKUSycvMuWNks5ryURHgaJpZM4Mv9jJ>
> .
>
